File: CCcodes_python/driverFULL_annual.py
import numpy as np
import pandas as pd
import scipy.io
import os

# Set display format for numpy arrays
np.set_printoptions(precision=15, floatmode='maxprec')

# Helper Functions ============================================================
def getData_annual(TT, iy, TIME):
    """Extract annual data for all variables and return as DataFrame"""
    TIME_annual = pd.DatetimeIndex(TIME[iy])
    annual_data = {}
    
    vars_to_process = [
        ('TAIR', 'TAIR_annual'), ('DWPT', 'DWPT_annual'),
        ('TAIRx', 'TAIRx_annual'), ('TAIRn', 'TAIRn_annual'),
        ('PRCP', 'PRCP_annual'), ('RELH', 'RELH_annual'),
        ('PRES', 'PRES_annual'), ('SM02', 'SM02_annual'),
        ('WDIR', 'WDIR_annual'), ('WSPD', 'WSPD_annual'),
        ('WSMX', 'WSMX_annual'), ('SRAD', 'SRAD_annual')
    ]
    
    for var_src, var_dest in vars_to_process:
        try:
            var_array = TT[var_src][iy] if TT[var_src].ndim > 1 else TT[var_src][iy]
            annual_data[var_dest] = var_array.flatten()
        except KeyError:
            raise ValueError(f"Missing variable {var_src} in dataset")

    return pd.DataFrame({
        'TIME_annual': TIME_annual,
        **annual_data
    }).set_index('TIME_annual').rename(columns=lambda x: x.replace('_annual', ''))

def getData_month(TT, mo, TIME):
    """Extract monthly data for all variables and return as DataFrame"""
    TIME_month = pd.DatetimeIndex(TIME[mo])
    
    vars_to_process = [
        ('TAIR', 'TAIR_month'), ('DWPT', 'DWPT_month'),
        ('TAIRx', 'TAIRx_month'), ('TAIRn', 'TAIRn_month'),
        ('PRCP', 'PRCP_month'), ('RELH', 'RELH_month'),
        ('PRES', 'PRES_month'), ('SM02', 'SM02_month'),
        ('WDIR', 'WDIR_month'), ('WSPD', 'WSPD_month'),
        ('WSMX', 'WSMX_month'), ('SRAD', 'SRAD_month')
    ]
    
    monthly_data = {}
    for var_src, var_dest in vars_to_process:
        try:
            var_array = TT[var_src][mo] if TT[var_src].ndim > 1 else TT[var_src][mo]
            monthly_data[var_dest] = var_array.flatten()
        except KeyError:
            raise ValueError(f"Missing required variable {var_src} in TT dataset")
    
    return pd.DataFrame({
        'TIME_month': TIME_month,
        **monthly_data
    }).set_index('TIME_month').rename(columns=lambda x: x.replace('_month', ''))

def calculate_run_length(arr):
    """Calculate maximum consecutive True values in a boolean array"""
    max_count = count = 0
    for val in arr:
        count = count + 1 if val else 0
        max_count = max(max_count, count)
    return max_count

def calculate_gdd(data, years, time_index, isD, ieD):
    """Calculate growing degree days for GD4 and GD10"""
    nm = 12
    gdd4 = np.full((len(years), nm), np.nan)
    gdd10 = np.full((len(years), nm), np.nan)
    
    for y_idx, year in enumerate(years):
        year_mask = (time_index.year == year)
        for m_idx in range(nm):
            mask = year_mask & (time_index.month == m_idx+1)
            if not mask.any(): continue
                
            try:
                tair = data['TAIR'][isD:ieD+1][mask]
                gdd4[y_idx, m_idx] = np.nansum(np.maximum(tair - 4, 0))
                gdd10[y_idx, m_idx] = np.nansum(np.maximum(tair - 10, 0))
            except KeyError:
                print(f"Missing TAIR data for {year}-{m_idx+1}")
                
    return {'GD4': gdd4, 'GD10': gdd10}

# Main Processing Function ====================================================
def main_processing():
    # Configuration
    site = 'BMNT'
    data_root = '/Volumes/Mesonet/winter_break/CCdata/'
    
    try:
        # Load data
        mesonet = scipy.io.loadmat(os.path.join(data_root, 'BMTN/01-Mar-2014_01-Aug-2023_BMTN_daily.mat'))
        time_data = pd.DatetimeIndex(mesonet['TimestampCollected'].flatten())
        
        # Date range calculations
        date_start = time_data[0] + pd.DateOffset(years=1)
        date_end = time_data[-1] - pd.DateOffset(years=1)
        isD = max(np.searchsorted(time_data, date_start.normalize(), side='left'), 0)
        ieD = min(np.searchsorted(time_data, date_end.normalize(), side='right'), len(time_data)-1)

        # Data validation
        try:
            TT = mesonet['TT_dailyMES'][isD:ieD+1]
        except KeyError:
            raise RuntimeError("Required dataset TT_dailyMES not found in MAT file")

        # Initialize output structure
        results = {
            'metadata': {
                'site': site,
                'start_date': time_data[isD],
                'end_date': time_data[ieD],
                'variables': [
                    'TAIR', 'DWPT', 'TAIRx', 'TAIRn', 'PRCP', 'RELH',
                    'PRES', 'SM02', 'WDIR', 'WSPD', 'WSMX', 'SRAD'
                ]
            },
            'years': np.unique(time_data[isD:ieD+1].year),
            'monthly': None,
            'annual': [],
            'indices': {},
            'temp_indices': {},
            'gdd': None
        }

        ny = len(results['years'])
        nm = 12

        # Initialize matrices
        FDm = np.zeros((ny, nm))
        CFDm = np.zeros((ny, nm))
        HDDm = np.zeros((ny, nm))
        IDm = np.zeros((ny, nm))
        TXnm = np.full((ny, nm), np.nan)
        TNnm = np.full((ny, nm), np.nan)

        # Process annual and monthly data
        monthly_data = np.full((ny, nm, len(results['metadata']['variables'])), np.nan)
        
        for y_idx, year in enumerate(results['years']):
            year_mask = (time_data[isD:ieD+1].year == year)
            
            try:
                # Annual data processing
                df_year = getData_annual(TT, year_mask, time_data)
                results['annual'].append(df_year)
                
                # Monthly aggregates
                monthly_means = df_year.groupby(df_year.index.month).mean()
                monthly_data[y_idx] = monthly_means.values.T

                # Temperature indices calculation
                for m_idx in range(nm):
                    time_mask = year_mask & (time_data[isD:ieD+1].month == m_idx+1)
                    
                    TAIR = TT['TAIR'][time_mask]
                    TAIRn = TT['TAIRn'][time_mask]
                    TAIRx = TT['TAIRx'][time_mask]

                    # Temperature extremes
                    TXnm[y_idx, m_idx] = np.nanmin(TAIRx)
                    TNnm[y_idx, m_idx] = np.nanmin(TAIRn)
                    
                    # Thermal indices
                    FDm[y_idx, m_idx] = np.nansum(TAIRn < 0)
                    frost_mask = TAIRn < 0
                    CFDm[y_idx, m_idx] = calculate_run_length(frost_mask) if np.any(frost_mask) else 0
                    HDDm[y_idx, m_idx] = np.nansum(18.3 - TAIR)
                    IDm[y_idx, m_idx] = np.nansum(TAIRx < 0)

            except Exception as e:
                print(f"Error processing {year}: {str(e)}")
                continue

        # Store results
        results['monthly'] = monthly_data
        results['gdd'] = calculate_gdd(TT, results['years'], time_data[isD:ieD+1], isD, ieD)
        results['temp_indices'] = {
            'FD': FDm, 'CFD': CFDm, 'HDD': HDDm, 'ID': IDm,
            'TXn': TXnm, 'TNn': TNnm,
            'years': results['years'], 'months': np.arange(1, 13)
        }

        print(f"Processing completed for {len(results['years'])} years")
        return results

    except Exception as e:
        print(f"Critical error in main processing: {str(e)}")
        raise

if __name__ == "__main__":
    analysis_results = main_processing()

File: CCcodes/driverFULL_annual.m
clear
format long

site='BMTN';

fileMES1='/BMTN/01-Mar-2014_01-Aug-2023_BMTN_daily.mat';
%/Volumes/Mesonet/winter_break/CCdata/BMTN/01-Mar-2014_01-Aug-2023_BMTN_daily.mat
%load Mesonet, Climate Data and Thresholds
dircli1='/Volumes/Mesonet/winter_break/CCdata/';
dircli=strcat(strcat(dircli1,site,'/'));
filethresh=strcat(dircli,strcat(site,"_CLIthresh_daily.mat"));
load(filethresh);
filemes=strcat(dircli1,fileMES1);
load(filemes);

TIME_full=TT_dailyMES.TimestampCollected;
sTIME_full=TIME_full(1);
eTIME_full=TIME_full(end);
YEAR_full=unique(TIME_full.Year);
[sYf,sMf,sDf]=ymd(TIME_full(1));
[eYf,eMf,eDf]=ymd(TIME_full(end));
dateS=datetime(sYf+1,01,01);
if dateS >= sTIME_full
   isD=find(all(ismember(TIME_full,dateS),2));
end
dateE=datetime(eYf-1,12,31); 
if dateE <= eTIME_full
   ieD=find(all(ismember(TIME_`full,dateE),2));
end

TT=TT_dailyMES(isD:ieD,:);
TIME=TT.TimestampCollected;

VAR=["TAIR_month" "DWPT_month" "TAIRx_month" "TAIRn_month" "PRCP_month" "RELH_month" ...
     "PRES_month" "SM02_month" "WDIR_month" "WSPD_month" "WSMX_month" "SRAD_month"]';
YEAR=unique(TIME.Year);

myStructA.var=VAR;
myStructA.year=YEAR;

ny=numel(YEAR);
nv=numel(VAR);

Md=cell(ny,nv);
My=cell(ny,nv);

yoD = year(TIME);         
moD = month(TIME);   

for i=1:ny
    k=2009+i;
    iy = ismember(yoD,k);
    TIME_annual=TIME(iy,:);
    for j=1:nv
        TTnew=getData_annual(TT,iy,TIME);
        TTcal=TIME;
        TTyears = year(TTcal);
        [~,TTindYears] = findgroups(TTyears);
        Vartt=TTnew{:,j};
        TTy{:,i}=TTnew;
        Ad{i,j}=TTy;
        Ay{i,j}=TTindYears;
    end
end

myStructA.year=Ay;
myStructA.data=Ad;

DATA=myStructA.data;

IndicesY=ones(ny,1);
for i=1:ny
    yearT(i)=datetime(double(YEAR(i)),01,01);
    IndicesY(i)=datefind(yearT(i),TIME_full);
end

%%%%%%%%%%%%%%%%%%function for annual data%%%%%%%%%%%%%%%%%%%%%
function [TT_annual]=getData_annual(TT,iy,TIME)

TIME_annual=TIME(iy);

TAIR_A=TT.TAIR;
TAIR_annual=TAIR_A(iy,:);
TT_TAIR_an=timetable(TIME_annual,TAIR_annual);

DWPT_A=TT.DWPT;
DWPT_annual=DWPT_A(iy,:);
TT_DWPT_an=timetable(TIME_annual,DWPT_annual);

TAIRx_A=TT.TAIRx;
TAIRx_annual=TAIRx_A(iy,:);
TTx_TAIR_an=timetable(TIME_annual,TAIRx_annual);

TAIRn_A=TT.TAIRn;
TAIRn_annual=TAIRn_A(iy,:);
TTn_TAIR_an=timetable(TIME_annual,TAIRn_annual);

PRCP_A=TT.PRCP;
PRCP_annual=PRCP_A(iy,:);
TT_PRCP_an=timetable(TIME_annual,PRCP_annual);

RELH_A=TT.RELH;
RELH_annual=RELH_A(iy,:);
TT_RELH_an=timetable(TIME_annual,RELH_annual);

PRES_A=TT.PRES;
PRES_annual=PRES_A(iy,:);
TT_PRES_an=timetable(TIME_annual,PRES_annual);

SM02_A=TT.SM02;
SM02_annual=SM02_A(iy,:);
TT_SM02_an=timetable(TIME_annual,SM02_annual);

WDIR_A=TT.WDIR;
WDIR_annual=WDIR_A(iy,:);
TT_WDIR_an=timetable(TIME_annual,WDIR_annual);

WSPD_A=TT.WSPD;
WSPD_annual=WSPD_A(iy,:);
TT_WSPD_an=timetable(TIME_annual,WSPD_annual);

WSMX_A=TT.WSMX;
WSMX_annual=WSMX_A(iy,:);
TT_WSMX_an=timetable(TIME_annual,WSMX_annual);

SRAD_A=TT.SRAD;
SRAD_annual=SRAD_A(iy,:);
TT_SRAD_an=timetable(TIME_annual,SRAD_annual);

TT_annual=[TT_TAIR_an TT_DWPT_an TTx_TAIR_an TTn_TAIR_an ...
    TT_PRCP_an TT_RELH_an TT_PRES_an TT_SM02_an TT_WDIR_an TT_WSPD_an TT_WSMX_an TT_SRAD_an];

end

%%%%%%%%%%%%%%%%%%%%%%%%INDEX CALCULATION%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%COLD INDICES%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%Growing Degree Days: GD4,GD10
GD4=zeros(nm,ny);
GD10=zeros(nm,ny);
for i=1:nm
    for j=1:ny
        TAIR=DATA{i,1}{1,j};
        GD4m(j,i)=sum(max((TAIR-4),0));
        GD10m(j,i)=sum(max((TAIR-10),0));
    end
end

%%%%%Growing Season Length (GSL)

%thresh=5;
%for i=1:1
%    TAIR=DATA{1,1}{1,i};
%    TairCalc=TAIR(IndicesY(i):(IndicesY(i+1)-IndicesY(i)));
%    TimeCalc=TIME(IndicesY(i):(IndicesY(i+1)-IndicesY(i)));
%    %start
%    [B,N,BI]=RunLength(TairCalc > thresh);
%    ix = find(N>thresh,1);
%    t1=TIME(ix+1);
%    %end
%    summerT=datetime(YEAR(i),07,01);
%    IndicesT=datefind(summerT,TimeCalc)-IndicesY(i)+1;
%    TIME2=TimeCalc(IndicesT:numel(TairCalc));
%    TAIR2=TairCalc(IndicesT:numel(TairCalc));
%    [B2,N2,BI2]=RunLength(TAIR2 < thresh);
%    %%%%%%%%%%%%%%%%%%%NEED TO FIX%%%%%%%%%%%%%%%%%
%    %if B2==1 & N2>thresh
%    %   ix2
%    %   %t2=TIME2(ix2+1);
%    %end
%    %GSL(i)=duration({t1;t2}); 
%end

%Frost Days: FD
for i=1:nm
    for j=1:ny
        TAIRn=DATA{i,5}{1,j};
        a1=find(TAIRn<0);
        FDm(i,j)=numel(a1);
    end
end

%Consecutive Frost Days: CFD
for i=1:nm
    for j=1:ny
        TAIRn=DATA{i,5}{1,j};
        [B,N,BI]=RunLength(TAIRn<0);
        CFDm(i,j)=max(double(B).*N,[],'omitnan');
    end
end

%Heating Degree Days: HDD
for i=1:nm
    for j=1:ny
        TAIR=DATA{i,1}{1,j};
        HDDm(i,j)=sum(18.3-TAIR,'omitnan');
    end
end

%Ice Days: ID
for i=1:nm
    for j=1:ny
        TAIRx=DATA{i,3}{1,j};
        a1=find(TAIRx<0);
        IDm(i,j)=numel(a1);
    end
end

%Cold Spell Duration Index: CSDI

%Cold Days: TG10p

%Cold Nights: TN10p

%Cold Day-Times: TX10p

%Minimum Value of Daily Maximum Temperature: TXn
for i=1:nm
    for j=1:ny
        TAIRx=DATA{i,3}{1,j};
        TXnm(i,j)=min(TAIRx,[],'omitnan');
    end
end

%Minimum Value of Daily Minimum Temperature: TNn
for i=1:nm
    for j=1:ny
        TAIRn=DATA{i,5}{1,j};
        TNnm(i,j)=min(TAIRn,[],'omitnan');
    end
end

%%%%%%%%%%%%%%%%%%%%%%COMPUND INDICES%%%%%%%%%%%%%%%%%%%%%%%%%%

%Cold Dry Days: CD

%Cold Wet Days: CW

%Warm Dry Days: WD

%Warm Wet Days: WW

%%%%%%%%%%%%%%%%%%%%%%DROUGHT INDICES%%%%%%%%%%%%%%%%%%%%%%%%%%

%Maximum Number of Consecutive Dry Days: CDD

%Three Month Standarized Precipitation Index: SPI3

%Six Month Standarized Precipitation Index: SPI6

%Potential Evapotranspiration: PET




%%%%%%%%%%%%%%%%%%function for monthly data%%%%%%%%%%%%%%%%%%%%%
function [TT_month]=getData_month(TT,mo,TIME)

TIME_month=TIME(mo,:);

TAIR_M=TT.TAIR;
TAIR_month=TAIR_M(mo,:);
TT_TAIR_mo=timetable(TIME_month,TAIR_month);

DWPT_M=TT.DWPT;
DWPT_month=DWPT_M(mo,:);
TT_DWPT_mo=timetable(TIME_month,DWPT_month);

TAIRx_M=TT.TAIRx;
TAIRx_month=TAIRx_M(mo,:);
TT_TAIRx_mo=timetable(TIME_month,TAIRx_month);

TAIRn_M=TT.TAIRn;
TAIRn_month=TAIR_M(mo,:);
TT_TAIRn_mo=timetable(TIME_month,TAIRn_month);

PRCP_M=TT.PRCP;
PRCP_month=PRCP_M(mo,:);
TT_PRCP_mo=timetable(TIME_month,PRCP_month);

RELH_M=TT.RELH;
RELH_month=RELH_M(mo,:);
TT_RELH_mo=timetable(TIME_month,RELH_month);

PRES_M=TT.PRES;
PRES_month=PRES_M(mo,:);
TT_PRES_mo=timetable(TIME_month,PRES_month);

SM02_M=TT.SM02;
SM02_month=SM02_M(mo,:);
TT_SM02_mo=timetable(TIME_month,SM02_month);

WDIR_M=TT.WDIR;
WDIR_month=WDIR_M(mo,:);
TT_WDIR_mo=timetable(TIME_month,WDIR_month);

WSPD_M=TT.WSPD;
WSPD_month=WSPD_M(mo,:);
TT_WSPD_mo=timetable(TIME_month,WSPD_month);

WSMX_M=TT.WSMX;
WSMX_month=WSMX_M(mo,:);
TT_WSMX_mo=timetable(TIME_month,WSMX_month);

SRAD_M=TT.SRAD;
SRAD_month=SRAD_M(mo,:);
TT_SRAD_mo=timetable(TIME_month,SRAD_month);

TT_month=[TT_TAIR_mo TT_DWPT_mo TT_TAIRx_mo TT_TAIRn_mo TT_PRCP_mo TT_RELH_mo TT_PRES_mo ...
          TT_SM02_mo TT_WDIR_mo TT_WSPD_mo TT_WSMX_mo TT_SRAD_mo];

end